### PR TITLE
Add onnxruntime test for new submitted models

### PR DESCRIPTION
**Pull request by jcwchen**
**Description**
- Use onnxruntime to test new submitted models in CIs.
- Consider exist test_data_set_0
- Design test_utils for reuse (used by test_all_models.py)

**Motivation**
Similar to https://github.com/onnx/models/pull/376
Improve model testing in CIs: test new models (submitted) by onnxruntime

### CODE REVIEW

File: .github/workflows/linux_ci.yml [hunk 1]
```diff
@@ -31,8 +31,19 @@ jobs:
...
           python -m pip install onnx==${{ matrix.onnx-version }} requests
-          python .azure-pipelines/test_models.py
\ No newline at end of file
+          python -m pip install onnxruntime
+          git clone https://github.com/microsoft/onnxruntime.git
+      - name: Test new models by onnx
+        run: |
+          export PYTHONPATH="$(pwd)/onnxruntime/tools/python/:$PYTHONPATH"
+          python workflow_scripts/test_models.py onnx
+      - name: Test new models by onnxruntime
```
Comment (jcwchen) [inline .github/workflows/linux_ci.yml:44]: I decoupled onnx and onnxruntime in CI. There might be other things we can decouple like testing the uploaded onnx model file and testing the model from uploaded .tar.gz

File: .github/workflows/windows_ci.yml [hunk 2]
```diff
@@ -31,8 +31,17 @@ jobs:
...
         run: |
           python -m pip install --upgrade pip
           python -m pip install onnx==${{ matrix.onnx-version }} requests
-          python .azure-pipelines/test_models.py
\ No newline at end of file
+          python -m pip install onnxruntime
+          git clone https://github.com/microsoft/onnxruntime.git
+      - name: Test new models by onnx
+        run: |
+          $Env:PYTHONPATH=".\onnxruntime\tools\python";
```
Comment (jcwchen) [inline .github/workflows/windows_ci.yml:42]: I cannot find a good way to set global env variables in Winodws-CI properly...  Use two `$Env:PYTHONPATH=".\onnxruntime\tools\python";` for now.

File: workflow_scripts/test_models.py [hunk 3]
```diff
@@ -1,48 +1,80 @@
...
-
-if len(failed_models) != 0:
-  print(str(len(failed_models)) +" models failed onnx checker.")
-  sys.exit(1)
-
-print(len(model_list), "model(s) checked.")
+
+def main(argv):
+  # default test by onnx and onnxruntime
+  test_onnx, test_onnxruntime = True, True
```
Comment (askhade) [inline workflow_scripts/test_models.py:11]: use argparse. Its very clean and let's you control optional, required ars, set defaults and much more.

File: workflow_scripts/test_models.py [hunk 4]
```diff
@@ -1,48 +1,80 @@
...
+        if os.path.exists(tar_gz_path):
+          test_utils.pull_lfs_file(tar_gz_path)
+          # check whether 'test_data_set_0' exists
+          model_path_from_tar, test_data_set = test_utils.extract_test_data(tar_gz_path)
+
+          # 1. check the onnx model from .tar.gz
+          if test_onnx: check_model.by_onnx(model_path_from_tar, tar_name)
+          if test_onnxruntime: check_model.by_onnxruntime(model_path_from_tar, tar_name, test_data_set)
+        else:
+          test_utils.pull_lfs_file(model_path)
```
Comment (askhade) [inline workflow_scripts/test_models.py:54]: should this be under else? Do you need this else? If onnx model is not downloaded then the code on line 57 won't run right?

Comment (jcwchen) [reply workflow_scripts/test_models.py:54]: Thank you for catching this bug. You are right.

File: workflow_scripts/check_model.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import onnx
+import test_utils
+
+
+def by_onnx(model_path, model_name):
+    model = onnx.load(model_path)
+    onnx.checker.check_model(model)
+    print('[PASS] {} is checked by onnx. '.format(model_name))
+
+def by_onnxruntime(model_path, model_name, test_data_set):
```
Comment (askhade) [inline workflow_scripts/check_model.py:12]: set default for test_data_set as None?

File: workflow_scripts/check_model.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import ort_test_dir_utils
+import onnxruntime
+import onnx
+import test_utils
+
+
+def by_onnx(model_path, model_name):
```
Comment (askhade) [inline workflow_scripts/check_model.py:7]: instead of by onnx maybe something like run_onnx_checker?

File: workflow_scripts/check_model.py (hunk 5 above)
Comment (askhade) [inline workflow_scripts/check_model.py:12]: instead of by_onnxruntime something like run_backend_ort ?

File: workflow_scripts/check_model.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+def by_onnxruntime(model_path, model_name, test_data_set):
+    # if 'test_data_set_N' doesn't exist, create test_dir
+    if not test_data_set:
+        onnxruntime.InferenceSession(model_path)
+        ort_test_dir_utils.create_test_dir(model_path, './', test_utils.TEST_ORT_DIR)
+        ort_test_dir_utils.run_test_dir(test_utils.TEST_ORT_DIR)
+    # otherwise use the existing 'test_data_set_N' as test data
+    else:
+        test_dir_from_tar = test_utils.get_model_directory(model_path)
+        ort_test_dir_utils.run_test_dir(test_dir_from_tar)
```
Comment (askhade) [inline workflow_scripts/check_model.py:21]: run_test_dir does not return an error in the case when expected outputs dont match actual output from the ort run. It simply prints a statement to stdout. This script won't catch this failure. You may have to update the run_test_dir script in ort to return a failure.

**Comment by jcwchen** [comment]
All corrected. @askhade Thank you for the helpful comments. For the uncaught failure in `ort_test_dir_utils`, I plan to clone that files directly and modify them on our own to make them meet our requirement. In that case, we don't need to change the original behavior and git clone ORT every time. Or do you prefer to do the update in ORT?

**Comment by askhade** [comment]
> All corrected. @askhade Thank you for the helpful comments. For the uncaught failure in `ort_test_dir_utils`, I plan to clone those files directly and modify them on our own to make them meet our requirement. In that case, we don't need to change the original behavior and git clone ORT every time. Or do you prefer to do the update in ORT?

I prefer updating ORT this way we are not duplicating utility code.

File: .github/workflows/windows_ci.yml (hunk 2 above)
Comment (askhade) [reply .github/workflows/windows_ci.yml:42]: Did you check this : https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables

File: workflow_scripts/test_models.py [hunk 8]
```diff
@@ -1,48 +1,82 @@
...
+  if target != 'onnxruntime':
+    check_model.run_onnx_checker(model_path, file_name)
+  if target != 'onnx':
+    check_model.run_backend_ort(model_path, file_name, test_data_set)
+
+def main():
+  parser = argparse.ArgumentParser(description='Test settings')
+  # default test by both onnx and onnxruntime
+  # if target is specified, only test by the specified one
+  parser.add_argument('--target', help='Test the model by which (onnx/onnxruntime)?')
```
Comment (askhade) [inline workflow_scripts/test_models.py:19]: nit: some other options required=False, default='None', type=str, choices=['onnx', 'onnxruntime', 'None'],

File: workflow_scripts/test_models.py [hunk 9]
```diff
@@ -1,48 +1,82 @@
...
-      print(e)
-      failed_models.append(model_path)
-
-if len(failed_models) != 0:
-  print(str(len(failed_models)) +" models failed onnx checker.")
-  sys.exit(1)
-
-print(len(model_list), "model(s) checked.")
+def check_by_target(model_path, file_name, test_data_set, target):
+  if target != 'onnxruntime':
```
Comment (askhade) [inline workflow_scripts/test_models.py:10]: why not target == 'onnxruntime' -> run onnxruntime check?

Comment (askhade) [reply workflow_scripts/test_models.py:10]: if you add choices value below then you can be sure target contains 1 of the 3 values provided

File: workflow_scripts/test_models.py [hunk 10]
```diff
@@ -1,48 +1,84 @@
...
+  for model_path in model_list:
+      model_name = model_path.split('/')[-1]
+      tar_name = model_name.replace('.onnx', tar_ext_name)
+      print('==============Testing {}=============='.format(model_name))
+
+      try:
+        # replace '.onnx' with '.tar.gz'
+        tar_gz_path = model_path[:-5] + '.tar.gz'
+        test_data_set = []
+        # if tar.gz exists, git pull and try to get test data
```
Comment (askhade) [inline workflow_scripts/test_models.py:49]: nit: This should be marked as step num 1

File: workflow_scripts/test_models.py [hunk 11]
```diff
@@ -1,48 +1,84 @@
...
+        if os.path.exists(tar_gz_path):
+          test_utils.pull_lfs_file(tar_gz_path)
+          # check whether 'test_data_set_0' exists
+          model_path_from_tar, test_data_set = test_utils.extract_test_data(tar_gz_path)
+
+          # 1. check the onnx model from .tar.gz
+          check_by_target(model_path_from_tar, tar_name, test_data_set, args.target)
+        
+        # git pull the onnx file
```
Comment (askhade) [inline workflow_scripts/test_models.py:58]: nit: this should be marked as step num 2

File: workflow_scripts/test_models.py [hunk 12]
```diff
@@ -1,48 +1,84 @@
...
+        test_utils.pull_lfs_file(model_path)
+
+        # 2. check the uploaded onnx model
+        check_by_target(model_path, model_name, test_data_set, args.target)
+        
+        if os.path.exists(tar_gz_path):
```
Comment (askhade) [inline workflow_scripts/test_models.py:64]: nit: you can get rid of this if else statement. Simply print success for ort run with tar folder after it is checked on line 57 and here simply add a success message for the onnx checker

File: workflow_scripts/test_models.py [hunk 13]
```diff
@@ -1,48 +1,84 @@
...
+      model_name = model_path.split('/')[-1]
+      tar_name = model_name.replace('.onnx', tar_ext_name)
+      print('==============Testing {}=============='.format(model_name))
+
+      try:
+        # replace '.onnx' with '.tar.gz'
+        tar_gz_path = model_path[:-5] + '.tar.gz'
+        test_data_set = []
+        # if tar.gz exists, git pull and try to get test data
+        if os.path.exists(tar_gz_path):
```
Comment (askhade) [inline workflow_scripts/test_models.py:50]: along with this if you can add one more check for target .i.e 
if os.path.exists(tar_gz_path) and (args.target == 'onnxruntime' or args.target == 'None'):
This way if target is only set to onnx then you dont need to even extract the tar file and then remove again .
If you do this then you dont need to call check_by_target, you can simply call run_backend_ort

File: workflow_scripts/test_models.py [hunk 14]
```diff
@@ -1,48 +1,84 @@
...
-    except Exception as e:
-      print(e)
-      failed_models.append(model_path)
-
-if len(failed_models) != 0:
-  print(str(len(failed_models)) +" models failed onnx checker.")
-  sys.exit(1)
-
-print(len(model_list), "model(s) checked.")
+def check_by_target(model_path, file_name, test_data_set, target):
```
Comment (askhade) [inline workflow_scripts/test_models.py:9]: I don't think this method is needed. SImply call the correct method to check from the parent script itself. With this method you are doing the check for target twice (once when you call this mtd for tar file to run with ort and second time when you call this mtd to run onnx checker)

File: workflow_scripts/test_models.py (hunk 13 above)
Comment (jcwchen) [reply workflow_scripts/test_models.py:50]: But I thought the onnx model inside the .tar file needs to be checked by onnx as well? Or we can simply assume it is   same as the .onnx file outside?

File: .github/workflows/windows_ci.yml (hunk 2 above)
Comment (jcwchen) [reply .github/workflows/windows_ci.yml]: Updated. Thanks for the info.

File: workflow_scripts/test_models.py [hunk 15]
```diff
@@ -1,48 +1,82 @@
...
-  print(str(len(failed_models)) +" models failed onnx checker.")
-  sys.exit(1)
-
-print(len(model_list), "model(s) checked.")
+
+def main():
+  parser = argparse.ArgumentParser(description='Test settings')
+  # default None: test by both onnx and onnxruntime
+  # if target is specified, only test by the specified one
+  parser.add_argument('--target', required=False, default='None', type=str,
```
Comment (askhade) [inline workflow_scripts/test_models.py:14]: nit: 'All' might be better than 'None'.

**Review by askhade — APPROVED** [review]
LGTM

**Comment by jcwchen** [comment]
:shipit:

**Review by wenbingl — APPROVED** [review]

File: .github/workflows/windows_ci.yml [hunk 16]
```diff
@@ -20,7 +20,9 @@ jobs:
         python-version: [3.7]
         architecture: ['x64']
         onnx-version: ['1.7']
-
+    # To get the tool ort_test_dir_utils from ORT
+    env:
+      PYTHONPATH: .\onnxruntime\tools\python
```
Comment (wenbingl) [inline .github/workflows/windows_ci.yml:25]: tools/python is not an official interface and is subject to change.
If we needs some python lines from this folder, we probably just copy it here. @jcwchen

Comment (jcwchen) [reply .github/workflows/windows_ci.yml:25]: I followed the instruction in the document to include the PYTHONPATH to use this PythonTools: https://github.com/microsoft/onnxruntime/blob/master/tools/python/PythonTools.md
I thought these files are more likely to change and actually it needs two files for now. It is possible that the ORT developers will introduce more files for this ORT utility. In that case, I suppose include the tools/python makes more sense here -- at least it is mentioned in the document and it should be maintained.

Comment (wenbingl) [reply .github/workflows/windows_ci.yml:25]: + @askhade for the discussion. My concern is these internal tools doesn't have backward compatibility and this model test will be easily broken after ONNXRuntime upgrading, while API in ORT python package are well defined and was promised to be backward compatible.

Comment (jcwchen) [reply .github/workflows/windows_ci.yml]: Removed the PYTHONPATH thing and uploaded the ort_test_dir_utils.py files. Thank you for the suggestion @wenbingl

**Pull request merged**